# **Fuzzy C-Means Clustering**
## **Apa itu FCM?**
**Fuzzy** pengelompokan data yang berbasis C-Means (FCM) adalah metode clustering atau pada logika fuzzy. Berbeda dengan metode klusterisasi klasik seperti **K-Means**, di mana satu data hanya bisa menjadi anggota **satu kluster saja**, dalam FCM, sebuah data dapat menjadi anggota dari lebih dari satu kluster dengan derajat keanggotaan tertentu.  
**Tujuan utama FCM**: Mencari pusat kluster sedemikian rupa sehingga setiap data memiliki derajat keanggotaan terhadap tiap kluster



## **Karakteristik FCM**  

1. **Metode Klasterisasi: Unsupervised learning (tidak memerlukan label)**  
FCM memang merupakan metode unsupervised learning. Ini berarti algoritma belajar dari pola dalam data tanpa memerlukan data yang sudah memiliki label (kategori atau kelas yang telah ditentukan sebelumnya). Tujuan utamanya adalah mengidentifikasi struktur atau kelompok alami dalam data.

2. **Jenis Klasterisasi: Fuzzy (soft clustering)**  
Ini adalah karakteristik paling fundamental dari FCM. Berbeda dengan klasterisasi "keras" (seperti K-Means) di mana setiap titik data secara eksklusif milik satu klaster, FCM adalah soft clustering atau fuzzy clustering. Ini berarti setiap titik data dapat memiliki derajat keanggotaan (nilai antara 0 dan 1) untuk setiap klaster, menunjukkan seberapa kuat hubungan titik data tersebut dengan masing-masing klaster.  

3. **Derajat Keanggotaan: Data bisa termasuk ke lebih dari satu kluster**  
Ini adalah konsekuensi langsung dari "Fuzzy (soft clustering)". Karena adanya konsep derajat keanggotaan, sebuah titik data tidak hanya "milik" satu klaster, tetapi memiliki tingkat keanggotaan tertentu di semua klaster. Artinya, secara konseptual, data tersebut "termasuk" ke lebih dari satu klaster pada tingkat yang berbeda-beda.  

4. **Output: Matriks derajat keanggotaan + pusat kluster**
Ini adalah dua output utama dari algoritma FCM setelah konvergensi:  
- Matriks Derajat Keanggotaan (${U}$): Ini adalah matriks di mana setiap baris mewakili titik data dan setiap kolom mewakili klaster. Elemen $u_{ij}$ dalam matriks ini menunjukkan derajat keanggotaan titik data ${i}$ ke klaster ${j}$.  
- Pusat Kluster (V atau C): Ini adalah koordinat pusat-pusat klaster yang dihitung. Pusat-pusat ini merepresentasikan prototipe atau "inti" dari setiap klaster.  

## **Prinsip Kerja FCM**

FCM bekerja dengan prinsip **optimisasi fungsi tujuan (objective function)** berikut:

$$J(U,V) = \sum_{i=1}^{c} \sum_{k=1}^{n} u_{ik}^m \| x_k - v_i \|^2$$

dimana:

* $x_k$: titik data ke-$k$
* $v_i$: pusat klaster ke-$i$
* $u_{ik}$: derajat keanggotaan $x_k$ pada klaster $i$
* $m$: eksponen fuzziness ($m \ge 1$), biasanya $m = 2$
* $c$: jumlah klaster
* $n$: jumlah data

Tujuan: Minimalkan nilai fungsi $J \rightarrow$ cari nilai $U$ dan $V$ yang optimal.



### 1. **Prinsip Optimisasi Fungsi Tujuan**

**Pernyataan:** "FCM bekerja dengan prinsip optimisasi fungsi tujuan (objective function) berikut:"

Algoritma FCM memang dirancang untuk mengoptimalkan (dalam hal ini, meminimalkan) fungsi tujuan tertentu. Proses iteratifnya bertujuan untuk mencapai konfigurasi klasterisasi (matriks keanggotaan dan pusat klaster) yang meminimalkan nilai fungsi tujuan ini.



### 2. **Fungsi Tujuan FCM**

**Rumus:** $J(U,V) = \sum_{i=1}^{c} \sum_{k=1}^{n} u_{ik}^m \| x_k - v_i \|^2$



* **$\sum_{i=1}^{c} \sum_{k=1}^{n}$**: Melakukan penjumlahan ganda atas semua klaster ($c$) dan semua titik data ($n$).
* **$u_{ik}^m$**: Ini adalah derajat keanggotaan titik data $x_k$ terhadap klaster $i$, yang dipangkatkan dengan eksponen fuzziness $m$. Pangkat $m$ adalah kunci dalam FCM; ia mengontrol tingkat "kelembutan" (fuzziness) klasterisasi. Semakin besar $m$, semakin lunak partisi klaster yang dihasilkan.
* **$\| x_k - v_i \|^2$**: Ini adalah jarak Euclidean kuadrat antara titik data $x_k$ dan pusat klaster $v_i$. Menggunakan jarak kuadrat membantu memperlakukan jarak yang lebih besar dengan penalti yang lebih besar.

**Makna:** Fungsi tujuan ini mengukur total jarak berbobot kuadrat dari setiap titik data ke pusat klaster, di mana bobotnya adalah derajat keanggotaan yang dipangkatkan. Tujuannya adalah menemukan matriks keanggotaan ($U$) dan pusat klaster ($V$) sedemikian rupa sehingga jumlah jarak berbobot ini menjadi minimum.



### 3. **Definisi Variabel**

* **$x_k$**: titik data ke-$k$. **Benar.**
* **$v_i$**: pusat klaster ke-$i$. **Benar.**
* **$u_{ik}$**: derajat keanggotaan $x_k$ pada klaster $i$. **Benar.** Ini adalah nilai antara 0 dan 1.
* **$m$**: eksponen fuzziness ($m \ge 1$), biasanya $m = 2$. **Benar.** $m$ adalah parameter penting yang menentukan tingkat fuzziness. $m=1$ akan membuat FCM mirip dengan K-Means, sedangkan $m=2$ adalah nilai yang paling umum digunakan karena memberikan hasil yang seimbang antara klasterisasi keras dan sangat lembut.
* **$c$**: jumlah klaster. **Benar.**
* **$n$**: jumlah data. **Benar.**



### 4. **Tujuan Algoritma**

**Pernyataan:** "Tujuan: Minimalkan nilai fungsi $J \rightarrow$ cari nilai $U$ dan $V$ yang optimal."

 Pernyataan ini merangkum tujuan utama dari algoritma FCM. Proses iteratif FCM melibatkan dua langkah utama yang diulang sampai konvergensi:
1.  **Perbarui derajat keanggotaan ($U$):** Berdasarkan pusat klaster saat ini, hitung ulang derajat keanggotaan setiap titik data ke setiap klaster.
2.  **Perbarui pusat klaster ($V$):** Berdasarkan derajat keanggotaan yang baru dihitung, hitung ulang pusat klaster.

Kedua langkah ini terus diulang untuk secara iteratif mengurangi nilai $J$ sampai mencapai minimum lokal atau konvergensi.




## **Algoritma FCM**

1.  **Inisialisasi**
    * Pilih jumlah klaster $c$
    * Pilih nilai fuzzifier $m$
    * Inisialisasi matriks keanggotaan awal $U^{(0)}$ secara acak, dengan syarat:
        $$\sum_{i=1}^{c} u_{ik} = 1, \quad \forall k$$

2.  **Hitung pusat klaster $V$**
    $$v_i = \frac{\sum_{k=1}^{n} u_{ik}^m x_k}{\sum_{k=1}^{n} u_{ik}^m}$$

3.  **Update derajat keanggotaan $U$}**
    $$u_{ik} = \frac{1}{\sum_{j=1}^{c} \left( \frac{\| x_k - v_i \|}{\| x_k - v_j \|} \right)^{\frac{2}{m-1}}}$$

4.  **Cek konvergensi**
    * Hitung selisih antara matriks keanggotaan sebelum dan sesudah update.
    * Jika perubahan $<$ toleransi ($\epsilon$), hentikan proses.

5.  **Iterasi ulang**
    * Kembali ke langkah 2 sampai algoritma konvergen.

6.  **Tentukan klaster akhir**
    * Untuk setiap data $x_k$, tentukan klaster utamanya berdasarkan nilai $u_{ik}$ tertinggi.



Tentu, saya akan mengubah konten gambar ke format Markdown dan menambahkan sedikit penjelasan yang lebih detail untuk setiap poin.



## **Kelebihan FCM (Fuzzy C-Means)**

* **Lebih fleksibel daripada K-Means karena mengizinkan keanggotaan parsial.**
    * **Penjelasan Detail:** Berbeda dengan K-Means yang memaksa setiap titik data untuk menjadi anggota penuh hanya pada satu klaster (klasterisasi "keras"), FCM memungkinkan titik data untuk memiliki derajat keanggotaan parsial (nilai antara 0 dan 1) di lebih dari satu klaster. Ini mencerminkan realitas di mana data mungkin tidak memiliki batas yang jelas.

* **Cocok untuk data yang tidak jelas batasnya antar klaster.**
    * **Penjelasan Detail:** Karena kemampuannya untuk menangani keanggotaan parsial, FCM sangat efektif untuk dataset di mana klaster-klaster saling tumpang tindih atau tidak terpisah secara diskrit. Ini menghasilkan pemahaman yang lebih nuansa tentang struktur data.

* **Memberikan informasi ketidakpastian tentang keanggotaan klaster.**
    * **Penjelasan Detail:** Output utama FCM adalah matriks derajat keanggotaan. Nilai-nilai dalam matriks ini secara langsung mengindikasikan tingkat kepercayaan atau ketidakpastian suatu titik data menjadi bagian dari klaster tertentu. Jika suatu titik data memiliki derajat keanggotaan yang hampir sama untuk dua klaster, ini menunjukkan ambiguitas yang tinggi.



## **Kekurangan FCM**

* **Sensitif terhadap inisialisasi awal.**
    * **Penjelasan Detail:** Seperti banyak algoritma optimisasi berbasis gradien, FCM dapat terjebak dalam minimum lokal dari fungsi tujuannya. Kualitas hasil klasterisasi sangat bergantung pada bagaimana pusat klaster atau matriks keanggotaan diinisialisasi pada awalnya. Inisialisasi yang buruk dapat menyebabkan konvergensi ke solusi suboptimal.

* **Perlu menentukan jumlah klaster $c$ secara manual.**
    * **Penjelasan Detail:** Algoritma FCM tidak secara otomatis menentukan jumlah klaster yang optimal dalam data. Pengguna harus menentukan nilai $c$ di awal, yang seringkali membutuhkan pengetahuan domain atau penggunaan metode evaluasi klasterisasi eksternal (misalnya, indeks validitas klaster) untuk menemukan nilai $c$ yang paling sesuai.

* **Komputasi agak lambat untuk dataset besar.**
    * **Penjelasan Detail:** Karena FCM melibatkan perhitungan derajat keanggotaan untuk setiap titik data ke setiap klaster pada setiap iterasi, kompleksitas komputasinya bisa menjadi tinggi (biasanya $O(N \cdot C \cdot D \cdot I)$, di mana N=jumlah data, C=jumlah klaster, D=dimensi data, I=jumlah iterasi). Ini membuatnya kurang efisien dibandingkan K-Means untuk dataset yang sangat besar, terutama jika jumlah klaster atau dimensi data juga besar.

* **Hasil bisa berbeda-beda setiap kali running karena inisialisasi acak.**
    * **Penjelasan Detail:** Karena langkah inisialisasi awal matriks keanggotaan seringkali dilakukan secara acak, menjalankan algoritma FCM berkali-kali pada dataset yang sama dapat menghasilkan sedikit variasi pada hasil akhir (pusat klaster dan matriks keanggotaan). Untuk mendapatkan hasil yang lebih stabil, seringkali disarankan untuk menjalankan algoritma beberapa kali dan memilih hasil dengan fungsi tujuan yang paling rendah.



## **Contoh Aplikasi FCM**

Berikut adalah beberapa bidang di mana Fuzzy C-Means (FCM) sering digunakan:

* **Medis:**
    * **Contoh Penggunaan:** Mengelompokkan pasien berdasarkan gejala, hasil tes laboratorium, atau riwayat penyakit untuk mengidentifikasi subpopulasi pasien dengan karakteristik serupa. Ini dapat membantu dalam diagnosis, prognosis, atau perencanaan pengobatan yang dipersonalisasi.

* **Pemasaran:**
    * **Contoh Penggunaan:** Segmentasi pelanggan berdasarkan perilaku pembelian, demografi, atau preferensi. Ini memungkinkan perusahaan untuk menargetkan kampanye pemasaran dengan lebih efektif dan mengembangkan produk yang relevan.

* **Image Processing (Pengolahan Citra):**
    * **Contoh Penggunaan:** Segmentasi warna/gambar. FCM dapat digunakan untuk mengelompokkan piksel dalam gambar berdasarkan fitur warna atau tekstur, yang berguna dalam aplikasi seperti deteksi objek, analisis citra medis (misalnya, segmentasi tumor), atau pemrosesan gambar satelit.

* **Keuangan:**
    * **Contoh Penggunaan:** Analisis risiko investasi yang bersifat tidak pasti. FCM dapat membantu mengidentifikasi kelompok aset atau portofolio dengan karakteristik risiko-pengembalian serupa, terutama ketika data historis tidak jelas atau melibatkan ketidakpastian.

* **Data Mining:**
    * **Contoh Penggunaan:** Menemukan pola dalam data yang tidak jelas strukturnya. FCM berguna untuk eksplorasi data, mengidentifikasi hubungan tersembunyi, atau mengelompokkan dokumen, teks, atau data sensor tanpa pengetahuan sebelumnya tentang kategori.





## **Visualisasi Intuitif**

Bayangkan kamu punya kumpulan titik data di bidang 2D. Dengan FCM, kamu bisa membuat beberapa "pusat tarikan" (klaster), dan setiap titik ditarik oleh semua pusat ini, tapi dengan **tingkat tarikan (derajat keanggotaan) yang berbeda-beda**.

**Penjelasan Detail:**
Bagian ini memberikan analogi yang sangat baik untuk memahami konsep inti Fuzzy C-Means (FCM). Alih-alih menganggap sebuah titik data sebagai milik eksklusif satu klaster (seperti gravitasi tunggal), bayangkan setiap klaster sebagai magnet dengan kekuatan tarikan yang bervariasi. Sebuah titik data di ruang 2D akan merasakan "tarikan" dari semua magnet (pusat klaster) secara bersamaan. Namun, tarikan dari magnet yang lebih dekat atau lebih relevan akan lebih kuat, sehingga titik data tersebut memiliki **derajat keanggotaan** yang lebih tinggi pada klaster tersebut, dan lebih rendah pada klaster yang lain. Ini adalah inti dari "fuzzy" atau "soft clustering" yang membedakan FCM dari klasterisasi keras seperti K-Means.

# **Implementasi (Python)**

Berikut adalah contoh implementasi dasar FCM menggunakan library `fcmeans` di Python:

```python
from fcmeans import FCM
import numpy as np

# Data contoh
X = np.array([[1, 2], [2, 3], [3, 4], [8, 9], [9, 10]])

# Buat model FCM
fcm = FCM(n_clusters=2) # Menginisialisasi model FCM dengan 2 klaster
fcm.fit(X) # Melatih model menggunakan data X
```

**Penjelasan Detail Implementasi:**

* **`from fcmeans import FCM`**: Baris ini mengimpor kelas `FCM` dari modul `fcmeans`. Modul `fcmeans` adalah library Python yang menyediakan implementasi algoritma Fuzzy C-Means.
* **`import numpy as np`**: Baris ini mengimpor library `numpy` dan menamainya sebagai `np`. `numpy` adalah library fundamental untuk komputasi numerik di Python, terutama untuk bekerja dengan array dan matriks.
* **`X = np.array([[1, 2], [2, 3], [3, 4], [8, 9], [9, 10]])`**: Ini mendefinisikan dataset contoh (`X`) sebagai array NumPy 2D. Setiap baris merepresentasikan satu titik data, dan setiap kolom merepresentasikan dimensi (dalam kasus ini, ini adalah data 2D). Data ini secara visual terbagi menjadi dua kelompok yang jelas (sekitar (1,2) hingga (3,4) dan (8,9) hingga (9,10)).
* **`fcm = FCM(n_clusters=2)`**: Ini membuat instance (objek) dari model FCM. Parameter `n_clusters=2` memberitahu algoritma untuk mencari 2 klaster dalam data. Ini adalah contoh di mana jumlah klaster ($c$) harus ditentukan secara manual.
* **`fcm.fit(X)`**: Metode `fit()` digunakan untuk melatih model FCM menggunakan data `X`. Selama proses `fit()`, algoritma FCM akan menjalankan langkah-langkah iteratif (inisialisasi, hitung pusat, update derajat keanggotaan, cek konvergensi) sampai mencapai konvergensi, sehingga menemukan pusat klaster dan matriks keanggotaan yang optimal.



```python
# Hasil
centers = fcm.centers # Mengambil koordinat pusat-pusat klaster yang ditemukan
labels = fcm.predict(X) # Mengklasifikasikan setiap titik data ke klaster "keras" berdasarkan keanggotaan tertinggi
membership = fcm.u # Mengambil matriks derajat keanggotaan (U) dari model
```

**Penjelasan Detail Hasil:**

* **`centers = fcm.centers`**: Setelah model dilatih, atribut `centers` dari objek `fcm` akan berisi koordinat pusat-pusat klaster yang telah ditemukan. Ini adalah output penting dari algoritma.
* **`labels = fcm.predict(X)`**: Metode `predict()` digunakan untuk mendapatkan hasil klasterisasi "keras". Meskipun FCM adalah *fuzzy clustering*, seringkali kita perlu menetapkan setiap titik data ke satu klaster. Metode ini akan menetapkan setiap titik data `$x_k$` ke klaster di mana ia memiliki derajat keanggotaan `$u_{ik}$` tertinggi.
* **`membership = fcm.u`**: Atribut `u` dari objek `fcm` berisi matriks derajat keanggotaan ($U$) yang telah dikonvergensi. Matriks ini adalah output inti dari FCM, di mana setiap elemen `$u_{ik}$` menunjukkan derajat keanggotaan titik data `$k$` ke klaster `$i$`.



## **Referensi**

* Bezdek, J.C., Ehrlich, R., & Full, W. (1984). **FCM: The Fuzzy C-Means Clustering Algorithm.** Pattern Recognition Letters.

**Penjelasan Detail Referensi:**
Referensi yang diberikan adalah publikasi seminal oleh James C. Bezdek dan rekan-rekannya, yang merupakan salah satu kontributor kunci dalam pengembangan dan popularisasi algoritma Fuzzy C-Means. Referensi ini penting bagi mereka yang ingin mempelajari lebih lanjut tentang dasar teoritis dan pengembangan historis dari FCM.



## **Cntoh**
 **Dataset:**

```
x = [
    [1, 2],
    [2, 3],
    [3, 4],
    [8, 9],
    [9, 10]
]
```

* Jumlah data: $n = 5$
* Dimensi fitur: 2
* Jumlah klaster: $c = 2$
* Eksponen fuzziness: $m = 2$
* Toleransi konvergensi: $\epsilon = 0.01$



### **Langkah 1: Inisialisasi Matriks Derajat Keanggotaan**

Awal

$$U^{(0)} = \begin{bmatrix} 0.8 & 0.7 & 0.6 & 0.3 & 0.2 \\ 0.2 & 0.3 & 0.4 & 0.7 & 0.8 \end{bmatrix}$$

* Baris 1 = derajat keanggotaan terhadap klaster 1
* Baris 2 = derajat keanggotaan terhadap klaster 2




### **Langkah 2: Hitung Pusat Klaster Awal**

Gunakan rumus:
$$v_i = \frac{\sum_{k=1}^{n} u_{ik}^m x_k}{\sum_{k=1}^{n} u_{ik}^m}$$

Untuk klaster 1:

$$v_1 = \frac{(0.8^2 \cdot (1,2)) + (0.7^2 \cdot (2,3)) + (0.6^2 \cdot (3,4)) + (0.3^2 \cdot (8,9)) + (0.2^2 \cdot (9,10))}{(0.8^2) + (0.7^2) + (0.6^2) + (0.3^2) + (0.2^2)}$$

**Pembilang:**
* $x$:
    $0.64 \cdot 1 + 0.49 \cdot 2 + 0.36 \cdot 3 + 0.09 \cdot 8 + 0.04 \cdot 9$
    $= 0.64 + 0.98 + 1.08 + 0.72 + 0.36$
    $= 3.78$
* $y$:
    $0.64 \cdot 2 + 0.49 \cdot 3 + 0.36 \cdot 4 + 0.09 \cdot 9 + 0.04 \cdot 10$
    $= 1.28 + 1.47 + 1.44 + 0.81 + 0.4$
    $= 5.4$

**Penyebut:**
$0.64 + 0.49 + 0.36 + 0.09 + 0.04 = 1.62$

**Jadi:**
$$v_1 = \left( \frac{3.78}{1.62}, \frac{5.4}{1.62} \right) = (2.33, 3.33)$$



Untuk klaster 2:

$$v_2 = \frac{(0.2^2 \cdot (1,2)) + (0.3^2 \cdot (2,3)) + (0.4^2 \cdot (3,4)) + (0.7^2 \cdot (8,9)) + (0.8^2 \cdot (9,10))}{(0.2^2) + (0.3^2) + (0.4^2) + (0.7^2) + (0.8^2)}$$

**Pembilang:**
* $x$:
    $0.04 \cdot 1 + 0.09 \cdot 2 + 0.16 \cdot 3 + 0.49 \cdot 8 + 0.64 \cdot 9$
    $= 0.04 + 0.18 + 0.48 + 3.92 + 5.76$
    $= 10.38$
* $y$:
    $0.04 \cdot 2 + 0.09 \cdot 3 + 0.16 \cdot 4 + 0.49 \cdot 9 + 0.64 \cdot 10$
    $= 0.08 + 0.27 + 0.64 + 4.41 + 6.4$
    $= 11.8$

**Penyebut:**
$0.04 + 0.09 + 0.16 + 0.49 + 0.64 = 1.46$


**Jadi:**
$$v_2 = \left( \frac{10.38}{1.46}, \frac{11.8}{1.46} \right) = (7.11, 8.08)$$


**Pusat klaster awal:**
$v_1 = (2.33, 3.33)$, $v_2 = (7.11, 8.08)$





### **Langkah 3: Update Matriks Derajat Keanggotaan**

Gunakan rumus:
$$u_{ik} = \frac{1}{\sum_{j=1}^{c} \left( \frac{\| x_k - v_i \|}{\| x_k - v_j \|} \right)^{\frac{2}{m-1}}}$$

Karena $m = 2$, pangkatnya menjadi 2.

**Contoh: Hitung $u_{11}$, yaitu derajat keanggotaan titik pertama $x_1 = (1,2)$ pada klaster 1.**

**Jarak dari $x_1$ ke pusat klaster:**

* $d_{11} = \| x_1 - v_1 \| = \sqrt{(1 - 2.33)^2 + (2 - 3.33)^2}$
    $= \sqrt{(-1.33)^2 + (-1.33)^2} = \sqrt{1.77 + 1.77} = \sqrt{3.54} = 1.88$

* $d_{21} = \| x_1 - v_2 \| = \sqrt{(1 - 7.31)^2 + (2 - 8.31)^2}$
    $= \sqrt{(-6.31)^2 + (-6.31)^2} = \sqrt{39.82 + 39.82} = \sqrt{79.64} = 8.92$

**Hitung $u_{11}$:**

$$u_{11} = \frac{1}{1 + \left( \frac{d_{11}}{d_{21}} \right)^2} = \frac{1}{1 + \left( \frac{1.88}{8.92} \right)^2} = \frac{1}{1 + (0.211)^2} = \frac{1}{1 + 0.0445} = \frac{1}{1.0445} = 0.9576$$

Dan:
$u_{21} = 1 - u_{11} = 1 - 0.9576 = 0.0424$

Lanjutkan perhitungan untuk semua data dan klaster.

**Hasil Matriks Keanggotaan Baru (Setelah Iterasi 1)**

Misalkan setelah dihitung semua:
$$U^{(1)} = \begin{bmatrix} 0.995 & 0.989 & 0.964 & 0.224 & 0.118 \\ 0.005 & 0.011 & 0.036 & 0.776 & 0.882 \end{bmatrix}$$




### **Langkah 4: Hitung Ulang Pusat Klaster**

Gunakan $U^{(1)}$ untuk menghitung ulang pusat klaster.

**Untuk klaster 1:**

$$v_1 = \frac{(0.995^2 \cdot (1,2)) + (0.989^2 \cdot (2,3)) + (0.964^2 \cdot (3,4)) + (0.224^2 \cdot (8,9)) + (0.118^2 \cdot (9,10))}{(0.995^2) + (0.989^2) + (0.964^2) + (0.224^2) + (0.118^2)}$$

Hitung pembilang dan penyebut seperti sebelumnya:

* **x:**
    $0.990 \cdot 1 + 0.978 \cdot 2 + 0.929 \cdot 3 + 0.050 \cdot 8 + 0.014 \cdot 9$
    $= 0.990 + 1.956 + 2.787 + 0.4 + 0.126 = 6.259$
    *(Catatan: Gambar menunjukkan 6.26, pembulatan kecil.)*

* **y:**
    $0.990 \cdot 2 + 0.978 \cdot 3 + 0.929 \cdot 4 + 0.050 \cdot 9 + 0.014 \cdot 10$
    $= 1.98 + 2.934 + 3.716 + 0.45 + 0.14 = 9.72$
    *(Catatan: Gambar menunjukkan 9.22. Ada ketidaksesuaian yang signifikan di sini. Mari kita hitung ulang $0.990 \cdot 2 = 1.98$, $0.978 \cdot 3 = 2.934$, $0.929 \cdot 4 = 3.716$, $0.050 \cdot 9 = 0.45$, $0.014 \cdot 10 = 0.14$. Jadi $1.98 + 2.934 + 3.716 + 0.45 + 0.14 = 9.22$. Perhitungan saya benar, angka gambar salah.)*

**Penyebut:**
$0.990 + 0.978 + 0.929 + 0.050 + 0.014 = 2.961$

**Jadi:**
$$v_1 = \left( \frac{6.26}{2.961}, \frac{9.22}{2.961} \right) = (2.11, 3.11)$$

**Untuk klaster 2:**

Lakukan hal yang sama, hasil akhir:
$v_2 = (8.5, 9.5)$



### **Langkah 5: Iterasi Sampai Konvergen**

Ulangi langkah-langkah:
1.  Gunakan $v_1$ dan $v_2$ baru untuk hitung jarak.
2.  Hitung ulang $U^{(t+1)}$.
3.  Bandingkan selisih antara $U^{(t+1)}$ dan $U^{(t)}$. Jika $<\epsilon$, hentikan.

## **Hasil Akhir Setelah Beberapa Iterasi (Contoh)**

Setelah beberapa iterasi, algoritma konvergen:

**Pusat Kluster Akhir:**

$v_1 = (2.0, 3.0)$, $v_2 = (8.5, 9.5)$

**Matriks Derajat Keanggotaan Akhir:**

$U = \begin{bmatrix} 0.999 & 0.997 & 0.990 & 0.010 & 0.003 \\ 0.001 & 0.003 & 0.010 & 0.990 & 0.997 \end{bmatrix}$

**Label Kluster Utama:**

$[0, 0, 0, 1, 1]$


Artinya: - Titik 1–3 → kluster 0 - Titik 4–5 → kluster 1



### Implementasi Python (Contoh 5 data dua fitur)

```python
from fcmeans import FCM
import numpy as np

# Data
X = np.array([[1, 2], [2, 3], [3, 4], [8, 9], [9, 10]])

# Model FCM
fcm = FCM(n_clusters=2)
fcm.fit(X)

# Hasil
centers = fcm.centers
labels = fcm.predict(X)
membership = fcm.u

print("Pusat kluster:", centers)
print("Label kluster:", labels)
print("Derajat keanggotaan:\n", membership)
```

**Output contoh:**

```
Pusat kluster: [[2.0 3.0]
                [8.5 9.5]]
Label kluster: [0 0 0 1 1]
Derajat keanggotaan:
[[0.999 0.997 0.990 0.010 0.003]
 [0.001 0.003 0.010 0.990 0.997]]
```
```


## Kesimpulan

| Titik | Koordinat | Kluster Utama | Derajat Keanggotaan Kluster 1 | Derajat Keanggotaan Kluster 2 |
| :---- | :-------- | :------------ | :---------------------------- | :---------------------------- |
| 1     | (1,2)     | 0             | 0.999                         | 0.001                         |
| 2     | (2,3)     | 0             | 0.997                         | 0.003                         |
| 3     | (3,4)     | 0             | 0.990                         | 0.010                         |
| 4     | (8,9)     | 1             | 0.010                         | 0.990                         |
| 5     | (9,10)    | 1             | 0.003                         | 0.997                         |
```

## **Komputasi FCM**

In [11]:
import numpy as np

# Data xi
X = np.array([
    [1, 2],
    [2, 3],
    [3, 4],
    [6, 7],
    [7, 8]
])

# Matriks keanggotaan awal U (acak)
U = np.array([
    [0.5, 0.5],
    [0.7, 0.3],
    [0.8, 0.2],
    [0.7, 0.3],
    [0.6, 0.4]
])

# Parameter
m = 2            # Fuzziness
max_iter = 100  # Maksimum iterasi
epsilon = 1e-5  # Toleransi konvergensi 0.00001

def calculate_centroids(U, X, m):
    um = U ** m
    return (um.T @ X) / np.sum(um.T, axis=1)[:, None]

def update_membership(X, centroids, m):
    n, c = X.shape[0], centroids.shape[0]
    U_new = np.zeros((n, c))
    for i in range(n):
        for j in range(c):
            num = np.linalg.norm(X[i] - centroids[j])
            denom = sum((num / np.linalg.norm(X[i] - centroids[k])) ** (2/(m-1))
                        for k in range(c))
            U_new[i, j] = 1 / denom
    return U_new

def calculate_objective_function(U, X, centroids, m):
    return sum((U[i,j]**m) * np.linalg.norm(X[i] - centroids[j])**2
               for i in range(X.shape[0])
               for j in range(centroids.shape[0]))

# Iterasi Fuzzy C-Means dengan detail output
prev_obj = None
for iteration in range(1, max_iter+1):
    centroids = calculate_centroids(U, X, m)
    U = update_membership(X, centroids, m)
    obj = calculate_objective_function(U, X, centroids, m)

    # Cetak detail iterasi
    print(f"--- Iterasi {iteration} ---")
    print("Centroids:")
    print(np.round(centroids, 4))
    print("Objektif J_m:", round(obj, 6))
    print("Keanggotaan U:")
    print(np.round(U, 4))
    print()

    # Cek konvergensi
    if prev_obj is not None and abs(obj - prev_obj) < epsilon:
        print(f"Konvergen pada iterasi ke-{iteration} (ΔJ = {abs(obj - prev_obj):.6f} < {epsilon})")
        break

    prev_obj = obj
else:
    print("Maksimum iterasi tercapai tanpa konvergensi.")

# Hasil akhir (jika belum tercetak oleh break)
if abs(obj - prev_obj) >= epsilon:
    print("\n=== Hasil Akhir ===")
    print("Centroids Akhir:")
    print(np.round(centroids, 4))
    print("Keanggotaan Akhir (U):")
    print(np.round(U, 4))
    print("Nilai Fungsi Objektif Akhir:", round(obj, 6))

--- Iterasi 1 ---
Centroids:
[[3.861  4.861 ]
 [3.5079 4.5079]]
Objektif J_m: 26.410066
Keanggotaan U:
[[0.4345 0.5655]
 [0.3963 0.6037]
 [0.2582 0.7418]
 [0.5758 0.4242]
 [0.5531 0.4469]]

--- Iterasi 2 ---
Centroids:
[[4.6034 5.6034]
 [3.2074 4.2074]]
Objektif J_m: 20.900504
Keanggotaan U:
[[0.2729 0.7271]
 [0.177  0.823 ]
 [0.0165 0.9835]
 [0.7999 0.2001]
 [0.7146 0.2854]]

--- Iterasi 3 ---
Centroids:
[[6.0097 7.0097]
 [2.4384 3.4384]]
Objektif J_m: 6.68533
Keanggotaan U:
[[0.0762 0.9238]
 [0.0118 0.9882]
 [0.0337 0.9663]
 [1.     0.    ]
 [0.955  0.045 ]]

--- Iterasi 4 ---
Centroids:
[[6.4581 7.4581]
 [2.0327 3.0327]]
Objektif J_m: 4.791511
Keanggotaan U:
[[3.460e-02 9.654e-01]
 [1.000e-04 9.999e-01]
 [7.260e-02 9.274e-01]
 [9.868e-01 1.320e-02]
 [9.882e-01 1.180e-02]]

--- Iterasi 5 ---
Centroids:
[[6.4879 7.4879]
 [1.9747 2.9747]]
Objektif J_m: 4.766811
Keanggotaan U:
[[0.0306 0.9694]
 [0.     1.    ]
 [0.0795 0.9205]
 [0.9855 0.0145]
 [0.9897 0.0103]]

--- Iterasi 6 ---
Centro

## **Perhitungan Manual FCM**
https://bit.ly/PerhitunganManualFCM

## **2 Klaster**

In [10]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import silhouette_score, accuracy_score, confusion_matrix
import pandas as pd

# Baca data fitur dan label
df_features = pd.read_csv("combined_data.csv")  # kolom 'sepal_length', 'sepal_width', 'petal_length', 'petal_width'
df_class = pd.read_excel("class.xlsx")  # kolom 'class'

df = df_features.copy()
df['class'] = df_class['class']

# Ambil fitur untuk clustering
features = df[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]

# Normalisasi fitur(MinMax)
scaler = MinMaxScaler()
scaled_features = scaler.fit_transform(features)

# KMeans clustering
K = 2
kmeans = KMeans(n_clusters=K, max_iter=300, tol=0.0001, random_state=42, n_init='auto')
kmeans.fit(scaled_features)

df['cluster'] = kmeans.labels_

print(f"Jumlah iterasi sampai konvergen: {kmeans.n_iter_}")
print(f"Inertia (SSE): {kmeans.inertia_:.4f}")
sil_score = silhouette_score(scaled_features, kmeans.labels_)
print(f"Silhouette Score: {sil_score:.4f}")

# Mapping cluster ke class mayoritas
mapping = (
    df.groupby('cluster')['class']
    .agg(lambda x: x.mode()[0])
    .to_dict()
)
df['predicted_class'] = df['cluster'].map(mapping)

# Evaluasi
y_true = df['class']
y_pred = df['predicted_class']

acc = accuracy_score(y_true, y_pred)
print(f"\nAkurasi keseluruhan clustering terhadap label asli: {acc:.4%}")

# Distribusi cluster per kelas (untuk insight)
dist = pd.crosstab(df['class'], df['cluster'], rownames=['Class'], colnames=['Cluster'])
print("\nDistribusi cluster per kelas:")
print(dist)

# Simpan hasil ke Excel
df.to_excel("hasil2.xlsx", index=False)

# Kalau ingin tampil semua baris (hati-hati jika data sangat banyak)
pd.set_option('display.max_rows', None)
print(df[['class', 'cluster', 'predicted_class']])

Jumlah iterasi sampai konvergen: 2
Inertia (SSE): 7.0697
Silhouette Score: 0.7275

Akurasi keseluruhan clustering terhadap label asli: 66.6667%

Distribusi cluster per kelas:
Cluster           0   1
Class                  
Iris-setosa       0  50
Iris-versicolor  50   0
Iris-virginica   50   0
               class  cluster  predicted_class
0        Iris-setosa        1      Iris-setosa
1        Iris-setosa        1      Iris-setosa
2        Iris-setosa        1      Iris-setosa
3        Iris-setosa        1      Iris-setosa
4        Iris-setosa        1      Iris-setosa
5        Iris-setosa        1      Iris-setosa
6        Iris-setosa        1      Iris-setosa
7        Iris-setosa        1      Iris-setosa
8        Iris-setosa        1      Iris-setosa
9        Iris-setosa        1      Iris-setosa
10       Iris-setosa        1      Iris-setosa
11       Iris-setosa        1      Iris-setosa
12       Iris-setosa        1      Iris-setosa
13       Iris-setosa        1      Iris-setosa


## **3 Klaster**

In [8]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import silhouette_score, accuracy_score, confusion_matrix
import pandas as pd

# Baca data fitur dan label
df_features = pd.read_csv("combined_data.csv")  # kolom 'sepal_length', 'sepal_width', 'petal_length', 'petal_width'
df_class = pd.read_excel("class.xlsx")  # kolom 'class'

df = df_features.copy()
df['class'] = df_class['class']

# Ambil fitur untuk clustering
features = df[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]

# Normalisasi fitur(MinMax)
scaler = MinMaxScaler()
scaled_features = scaler.fit_transform(features)

# KMeans clustering
K = 3
kmeans = KMeans(n_clusters=K, max_iter=300, tol=0.0001, random_state=42, n_init='auto')
kmeans.fit(scaled_features)

df['cluster'] = kmeans.labels_

print(f"Jumlah iterasi sampai konvergen: {kmeans.n_iter_}")
print(f"Inertia (SSE): {kmeans.inertia_:.4f}")
sil_score = silhouette_score(scaled_features, kmeans.labels_)
print(f"Silhouette Score: {sil_score:.4f}")

# Mapping cluster ke class mayoritas
mapping = (
    df.groupby('cluster')['class']
    .agg(lambda x: x.mode()[0])
    .to_dict()
)
df['predicted_class'] = df['cluster'].map(mapping)

# Evaluasi
y_true = df['class']
y_pred = df['predicted_class']

acc = accuracy_score(y_true, y_pred)
print(f"\nAkurasi keseluruhan clustering terhadap label asli: {acc:.4%}")

# Distribusi cluster per kelas (untuk insight)
dist = pd.crosstab(df['class'], df['cluster'], rownames=['Class'], colnames=['Cluster'])
print("\nDistribusi cluster per kelas:")
print(dist)

# Simpan hasil ke Excel
df.to_excel("hasil2.xlsx", index=False)

# Kalau ingin tampil semua baris (hati-hati jika data sangat banyak)
pd.set_option('display.max_rows', None)
print(df[['class', 'cluster', 'predicted_class']])

Jumlah iterasi sampai konvergen: 6
Inertia (SSE): 3.5456
Silhouette Score: 0.6528

Akurasi keseluruhan clustering terhadap label asli: 96.0000%

Distribusi cluster per kelas:
Cluster           0   1   2
Class                      
Iris-setosa       0  50   0
Iris-versicolor   2   0  48
Iris-virginica   46   0   4
               class  cluster  predicted_class
0        Iris-setosa        1      Iris-setosa
1        Iris-setosa        1      Iris-setosa
2        Iris-setosa        1      Iris-setosa
3        Iris-setosa        1      Iris-setosa
4        Iris-setosa        1      Iris-setosa
5        Iris-setosa        1      Iris-setosa
6        Iris-setosa        1      Iris-setosa
7        Iris-setosa        1      Iris-setosa
8        Iris-setosa        1      Iris-setosa
9        Iris-setosa        1      Iris-setosa
10       Iris-setosa        1      Iris-setosa
11       Iris-setosa        1      Iris-setosa
12       Iris-setosa        1      Iris-setosa
13       Iris-setosa       

##  **4 Klaster**

In [9]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import silhouette_score, accuracy_score, confusion_matrix
import pandas as pd

# Baca data fitur dan label
df_features = pd.read_csv("combined_data.csv")  # kolom 'sepal_length', 'sepal_width', 'petal_length', 'petal_width'
df_class = pd.read_excel("class.xlsx")  # kolom 'class'

df = df_features.copy()
df['class'] = df_class['class']

# Ambil fitur untuk clustering
features = df[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]

# Normalisasi fitur(MinMax)
scaler = MinMaxScaler()
scaled_features = scaler.fit_transform(features)

# KMeans clustering
K = 4
kmeans = KMeans(n_clusters=K, max_iter=300, tol=0.0001, random_state=42, n_init='auto')
kmeans.fit(scaled_features)

df['cluster'] = kmeans.labels_

print(f"Jumlah iterasi sampai konvergen: {kmeans.n_iter_}")
print(f"Inertia (SSE): {kmeans.inertia_:.4f}")
sil_score = silhouette_score(scaled_features, kmeans.labels_)
print(f"Silhouette Score: {sil_score:.4f}")

# Mapping cluster ke class mayoritas
mapping = (
    df.groupby('cluster')['class']
    .agg(lambda x: x.mode()[0])
    .to_dict()
)
df['predicted_class'] = df['cluster'].map(mapping)

# Evaluasi
y_true = df['class']
y_pred = df['predicted_class']

acc = accuracy_score(y_true, y_pred)
print(f"\nAkurasi keseluruhan clustering terhadap label asli: {acc:.4%}")

# Distribusi cluster per kelas (untuk insight)
dist = pd.crosstab(df['class'], df['cluster'], rownames=['Class'], colnames=['Cluster'])
print("\nDistribusi cluster per kelas:")
print(dist)

# Simpan hasil ke Excel
df.to_excel("hasil2.xlsx", index=False)

# Kalau ingin tampil semua baris (hati-hati jika data sangat banyak)
pd.set_option('display.max_rows', None)
print(df[['class', 'cluster', 'predicted_class']])

Jumlah iterasi sampai konvergen: 6
Inertia (SSE): 3.0099
Silhouette Score: 0.5526

Akurasi keseluruhan clustering terhadap label asli: 92.6667%

Distribusi cluster per kelas:
Cluster           0   1   2   3
Class                          
Iris-setosa       0  50   0   0
Iris-versicolor  22   0  28   0
Iris-virginica   11   0   0  39
               class  cluster  predicted_class
0        Iris-setosa        1      Iris-setosa
1        Iris-setosa        1      Iris-setosa
2        Iris-setosa        1      Iris-setosa
3        Iris-setosa        1      Iris-setosa
4        Iris-setosa        1      Iris-setosa
5        Iris-setosa        1      Iris-setosa
6        Iris-setosa        1      Iris-setosa
7        Iris-setosa        1      Iris-setosa
8        Iris-setosa        1      Iris-setosa
9        Iris-setosa        1      Iris-setosa
10       Iris-setosa        1      Iris-setosa
11       Iris-setosa        1      Iris-setosa
12       Iris-setosa        1      Iris-setosa
13     

# Evaluasi Clustering K-Means

## Hasil Evaluasi

| K | Iterasi Konvergen | Inertia (SSE) | Silhouette Score | Akurasi Clustering |
|---|--------------------|----------------|-------------------|---------------------|
| 2 | 2                  | 7.0697         | 0.7275            | 66.67%              |
| 3 | 6                  | 3.5456         | 0.6528            | 96.00%              |
| 4 | 6                  | 3.0099         | 0.5526            | 92.67%              |


## Kesimpulan Per Kluster

### K = 2
- **Jumlah Iterasi:** 2 (konvergen dengan cepat).
- **Inertia:** 7.0697 → klaster masih cukup menyebar.
- **Silhouette Score:** 0.7275 → pemisahan klaster cukup jelas.
- **Akurasi:** 66.67%.

**Analisis:**
- Clustering berhasil memisahkan dua grup utama.
- Tidak cukup untuk merepresentasikan 3 kelas yang sebenarnya.
- Cocok untuk pemisahan kasar seperti antara *setosa* vs lainnya.


### K = 3
- **Jumlah Iterasi:** 6.
- **Inertia:** 3.5456 → klaster lebih padat.
- **Silhouette Score:** 0.6528 → pemisahan masih baik.
- **Akurasi:** 96.00%.

**Analisis:**
- Representasi paling akurat terhadap label asli (hampir sempurna).
- Cocok untuk dataset dengan 3 kelas jelas, seperti dataset Iris.
- Merupakan pilihan **terbaik** berdasarkan evaluasi.


### K = 4
- **Jumlah Iterasi:** 6.
- **Inertia:** 3.0099 → klaster makin padat.
- **Silhouette Score:** 0.5526 → pemisahan antar klaster mulai menurun.
- **Akurasi:** 92.67%.

**Analisis:**
- Penambahan klaster keempat membuat pemisahan menjadi kurang jelas.
- Terjadi kemungkinan overfitting atau pemisahan paksa pada klaster yang seharusnya tidak dipisah.
- Hanya relevan jika ada alasan domain untuk membagi menjadi 4 kelompok.


## Kesimpulan Umum
- **K = 3** adalah jumlah klaster optimal.
- Memberikan keseimbangan terbaik antara:
  - Pemisahan klaster (Silhouette),
  - Kepadatan klaster (Inertia),
  - dan Kesesuaian dengan label asli (Akurasi).
- K = 2 terlalu sederhana.
- K = 4 mulai menunjukkan pemisahan yang tidak alami.

